File: LeetCode problem set/21_Merge_Two_Sorted_Lists.cpp
// https://leetcode.com/problems/merge-two-sorted-lists/
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
      if(!l1) return l2;
      if(!l2) return l1;

      // selecting small value as head
      ListNode* head = l1;
      if(l1->val > l2->val){
        head = l2;
        l2 = l2->next;
      } else {
        l1 = l1->next;
      }

      ListNode* curr = head;
      while(l1 && l2) {
        if(l1->val < l2->val) {
          curr->next = l1;
          l1 = l1->next;
        } else {
          curr->next = l2;
          l2 = l2->next;
        }
        curr = curr->next;
      }

      if(!l1) curr->next = l2;
      else curr->next = l1;

      return head;
    };
};

int main() {
  ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);
  Solution sol;
  ListNode* arr1 = new ListNode(1);
  arr1->next = new ListNode(2);
  arr1->next->next = new ListNode(4);
  ListNode* arr2 = new ListNode(1);
  arr2->next = new ListNode(3);
  arr2->next->next = new ListNode(4);
  sol.mergeTwoLists(arr1, arr2);
  return 0;
}

File: LeetCode problem set/21_Merge_Two_Sorted_Lists.js
// https://leetcode.com/problems/merge-two-sorted-lists/

const mergeTwoLists = (list1, list2) => {
  if (!list1) return list2;
  if (!list2) return list1;

  const mergedList = [];
  let idx1 = 0,
    idx2 = 0,
    idxT = 0;

  while (list1.length && list2.length) {
    if (list1[idx1] <= list2[idx2]) {
      mergedList[idxT] = list1[idx1];
      idx1++;
      idxT++;
    } else {
      mergedList[idxT] = list2[idx2];
      idx2++;
      idxT++;
    }
  }

  while (idx1 < list1.length) {
    mergedList[idxT] = list1[idx1];
    idx1++;
    idxT++;
  }

  while (idx2 < list2.length) {
    mergedList[idxT] = list2[idx2];
    idx2++;
    idxT++;
  }

  return mergedList;
};

const list1 = [1, 2, 4];
const list2 = [1, 3, 4];

const list = mergeTwoLists(list1, list2);
